spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/TableWriter.scala

```diff
@@ -97,42 +97,53 @@ class TableWriter[T] private (
     }
   }
 
+  def batchRoutingKey(session: Session, routingKeyGenerator: RoutingKeyGenerator)(bs: BoundStatement): Any = {
+    writeConf.batchLevel match {
+      case BatchLevel.All => 0
+
+      case BatchLevel.ReplicaSet =>
+        if (bs.getRoutingKey == null)
+          bs.setRoutingKey(routingKeyGenerator(bs))
+        session.getCluster.getMetadata.getReplicas(keyspaceName, bs.getRoutingKey).hashCode() // hash code is enough
+
+      case BatchLevel.Partition =>
+        if (bs.getRoutingKey == null) {
+          bs.setRoutingKey(routingKeyGenerator(bs))
+        }
+        bs.getRoutingKey.duplicate()
+    }
+  }
+
+  def newRateLimiter = {
+    if (!OutputMetricsUpdater.detailedMetricsEnabled && writeConf.throughputMBPS < WriteConf.DefaultThroughputMBPS)
+      logError(
```

Comment:
Maybe we should just have writeConf.throughputMBPS automatically enable detailed metrics? (Or enable metrics and warn that it needed to enable metrics?